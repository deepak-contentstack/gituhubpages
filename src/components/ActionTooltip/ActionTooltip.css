@import '../../../tokens/variables.scss';
@import '../../../tokens/tokens.css';
@import '../../static/styles/css/typography.css';
@import '../../static/styles/css/common.css';

.ActionTooltip {
    align-items: center;
    display: flex;
    font-family: $font-family-primary;
    position: relative;
    width: 100%;
    
    &:focus{
        background : $color-brand-white-darker;
    }

    &__actions {
        align-items: center;
        background-color: $color-base-white-5;
        border-radius: $radii-6;
        box-shadow: 0 $space-8 $space-20 rgba(34, 34, 34, 0.1);
        /* display: none; */
        height: $space-32;
        position: absolute;
        right: $space-20;
        width: auto;
        li {
            border-radius: 6px;
            cursor: pointer;
            &:focus{
                box-shadow: 0 0 0 1px #6c5ce7,0 0 0 4px #bbb4f4;
                position: relative;
                z-index: 1;
                .wrappper-svg{
                    background: $color-base-gray-5;
                    &> svg {
                        path {
                            fill: $color-base-black-5;
                        }
                    }
                }
            }
            .wrappper-svg{
                align-items: center;
                border-radius: $radii-6;
                cursor: pointer;
                display: flex;
                font-weight: $font-weight-extra-bold;
                height: $space-32;
                justify-content: center;
                width: $space-32;
                &:hover{
                    background: $color-base-gray-5;
                    &> svg {
                        path {
                            fill: $color-base-black-5;
                        }
                    }
                }
            }
        }
    }

    &:hover {
        &.ActionTooltip--primary {
            background-color: $color-brand-white-darker;
        }
    }

    &.ActionTooltip--primary {
        font-size: $size-font-medium;
        font-weight: $font-weight-medium;
    }
}

.ActionTooltip-row {
    height:  3.75rem;
}

.ActionTooltip:hover  .ActionTooltip__actions {
    display: flex;
}

.ActionTooltip--Select {
    &:hover {
        .Select {
            &:not(:hover,:focus) {
                .Select__control {
                    background-color: $color-border-lighter;
                    svg {
                        path[stroke], circle[stroke], rect[stroke] {
                            stroke: $color-base-gray-40;
                        }
                        path[fill], circle[fill], rect[fill] {
                            fill: $color-base-gray-40;
                        }
                        path + path {
                            fill: $color-brand-white-base !important;
                        }
                    }
                }
            }
        }
    }
}

.ActionTooltip--TextInput {
    &:hover {
        .TextInput__input {
            &:not(:hover, :focus) {
                background-color: $color-border-lighter;
            }
        }
    }
}