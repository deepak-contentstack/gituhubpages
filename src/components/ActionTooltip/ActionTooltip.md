## **ActionTooltip**

A ActionTooltip Component can be used to provide tooltip with set of operation the user can perform when hovered on certain list/ELement.

#### **Props**:

- **list**: Array of Objects

  - required
  - Structure: {
    label : string , component :JSX | React Node ,
    title : string
    action : function to trigger when clicked,
    }

- **type**?: `'primary' | 'secondary'`

  - Optional
  - Default value: `'primary'`
