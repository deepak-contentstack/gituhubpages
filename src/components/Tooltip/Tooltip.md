## **Tooltip**

A Tooltip Component can be used to indicate the user what is the functionality of certain element.

### **Tooltip Variant use**

- **dark**: Tooltip with dark variant can be used to show additional information for certain element.

  - props:
    - variant: dark

- **light**: Tooltip that has more details and components as a content (eg: Tooltip Card, Publish status.)

  - props:
    - variant: light

#### **Props**:

- **content**: String or React Component

  - required

- **position**: String

  - required
  - values : right | left | top | bottom

- **variantType**: String

  -optional (defaults to dark)
  -values : dark | light

- **showArrow**: Boolean

  - optional

- **trigger**: String

  - optional
  - default value : 'mouseenter focus'
  - other values: 'click', 'focusin', 'mouseenter click'

* **onHidden**: function
  Invoked once after tooltip hides.
  - optional


* **visible**: Boolean
  - Control the tooltips visibility
  - optional


* **showClose**: Boolean
  - Show a close icon at the top right corner of the tooltip
  - optional

* **onClose**: Functiom
  - Callback function to be executed after close icon is clicked
  - optional


