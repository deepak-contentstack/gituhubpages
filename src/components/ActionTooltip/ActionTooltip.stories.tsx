import React from 'react'
import { withDesign } from 'storybook-addon-designs'
// @ts-ignore
import notes from './ActionTooltip.md'
import ActionTooltip, { IActionTooltip } from './ActionTooltip'
// import Icon from '../Icon2/Icon'

const tableDisable = { table: { disable: true } }

export default {
  title: 'Components / ActionTooltip',
  component: ActionTooltip,
  decorators: [withDesign],
  parameters: {
    componentSubtitle: 'Operation tool',
    notes: { notes },
    design: {
      type: 'figma',
      url: 'https://www.figma.com/proto/ovlEsic8nr81y5BOu1OlqD/ui-inventory-v2?node-id=213%3A558&scaling=min-zoom'
    }
  },
  argTypes: {
    className: {
      control: { type: 'text' },
      description: 'Option for adding custom class',
      table: {
        type: {
          summary: 'string'
        }
      },
      type: {
        required: false
      }
    },
    testId: {
      control: { type: 'text' },
      description: 'An ID used for testing purposes applied as a data attribute (data-test-id)',
      table: {
        type: {
          summary: 'string'
        },
        defaultValue: {
          summary: 'cs-actiontooltip'
        }
      },
      type: {
        required: false
      }
    },
    type: {
      control: { type: 'radio', options: ['primary', 'secondary'] },
      description: 'Provide the type of tooltip ',
      table: {
        type: {
          summary: 'string'
        },
        defaultValue: {
          summary: 'primary'
        }
      },
      type: {
        required: false
      }
    },
    children: {
      control: { type: 'object' },
      description: 'Provide the content for which Tooltip needs to be displayed',
      table: {
        type: {
          summary: 'string|ReactNode'
        }
      }
    },
    list: {
      control: { type: 'object' },
      description: 'Provide the list of actions for the tooltip',
      table: {
        type: {
          summary: 'array'
        }
      }
    },
    data: {
      control: { type: 'disabled' },
      description: 'Provide the data for the component, if provided as label',
      table: {
        type: {
          summary: 'array'
        }
      }
    },
    label: {
      control: { type: 'disabled' },
      description: 'Provide the label for the tooltip as a component or string',
      table: {
        type: {
          summary: 'string|ReactNode'
        }
      }
    },
    right: {
      control: { type: 'text' },
      description: 'The right CSS property to specify the horizontal position of tooltip',
      table: {
        type: {
          summary: 'string|number'
        }
      }
    }
  }
}

export const Default = (args: IActionTooltip) => {
  const toolList = [
    {
      label: () => <div></div>,
      title: 'Edit',
      action: () => console.log('Edit triggered')
    },
    {
      label: <div></div>,
      title: 'Copy UID',
      action: () => console.log('Copy triggered'),
      displayType: 'clipboard',
      copyText: 'ct_uid'
    },
    {
      label: <div></div>,
      title: 'Publish',
      action: () => console.log('PublishIcon triggered')
    },
    {
      label: <div></div>,
      title: 'Trash',
      action: () => console.log('Trash triggered'),
      className: 'ActionListItem--warning'
    }
  ]

  return (
    <>
      <ActionTooltip list={toolList} className={args.className} type={args.type} right={args.right}>
        {args.children}
      </ActionTooltip>
    </>
  )
}

Default.args = {
  className: 'ActionTooltip-row',
  type: 'primary',
  children: 'Can Augmented And Future Reality Build Empathy?',
  testId: 'cs-actiontooltip',
  right: '0'
}

Default.argTypes = {
  list: tableDisable,
  data: tableDisable,
  label: tableDisable
}

export const CustomAction = (args) => {
  const actions = [
    {
      label: <div></div>,
      title: args.title,
      action: () => console.log(`${args.title} triggered`),
      className: args.className
    }
  ]

  return (
    <div className="flex-column">
      <span className="pb-20 font-size-medium"></span>
      <ActionTooltip list={actions} className="ActionTooltip-row">
        The {args.title} action for this row has a custom classname
      </ActionTooltip>
    </div>
  )
}

CustomAction.args = {
  className: 'ActionListItem--warning',
  title: 'Trash'
}

CustomAction.argTypes = {
  className: {
    control: { type: 'text' },
    description: 'Provide the custom className to action in the tooltip',
    table: {
      type: {
        summary: 'string'
      }
    }
  },
  title: {
    control: { type: 'text' },
    description: 'Provide the title/icon to action in the tooltip',
    table: {
      type: {
        summary: 'string'
      }
    }
  },
  list: tableDisable,
  data: tableDisable,
  right: tableDisable,
  label: tableDisable,
  testId: tableDisable,
  children: tableDisable,
  type: tableDisable
}
