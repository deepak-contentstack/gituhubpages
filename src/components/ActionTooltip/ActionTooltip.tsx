import React, { useState, useEffect, useRef } from 'react'
// import { useComponentVisible } from '../DropDown/Dropdown'
//import ClipBoard from '../ClipBoard/ClipBoard'
import Tooltip from '../Tooltip/Tooltip'
import './ActionTooltip.css'
import cn from 'classnames'
export type ActionTooltipList = {
  label: string | React.ReactNode
  title: string
  action: Function
  className?: string
}
export type IActionTooltip = {
  type?: 'primary' | 'secondary'
  list: Array<ActionTooltipList>
  children?: React.ReactNode | Element
  className?: string
  data?: any
  right?: string | number,
  testId?: string
}

function useComponentVisible(initialIsVisible: boolean) {
  const [isComponentVisible, setIsComponentVisible] = useState(initialIsVisible)
  const ref: any = useRef(null)

  const handleHideDropdown = (event: KeyboardEvent) => {
    if (event.key === 'Escape') {
      setIsComponentVisible(false)
    }
  }

  const handleClickOutside = (event: any) => {
    if ((ref.current && !ref.current.contains(event.target)) || (document.activeElement === document.getElementById('sidebar-extension')) || (document.activeElement.clientWidth == 315)) {
      setIsComponentVisible(false)
    }
  }

  useEffect(() => {
    document.addEventListener('keydown', handleHideDropdown, true)
    document.addEventListener('click', handleClickOutside, true)
    return () => {
      document.removeEventListener('keydown', handleHideDropdown, true)
      document.removeEventListener('click', handleClickOutside, true)
    }
  })

  return { ref, isComponentVisible, setIsComponentVisible }
}

const ActionTooltip: React.FunctionComponent<IActionTooltip> = (props) => {
  const { list, children, className, data, testId, type, ...rest } = props
  const { ref, isComponentVisible, setIsComponentVisible } = useComponentVisible(false)
  const classNames = cn('ActionTooltip', `ActionTooltip--${type}`, className)
  const RenderLabel = (props) => {
    const Component = props.component
    return <Component data={props.data} />
  }

  return (
    <div
      className={classNames}
      ref={ref}
      data-test-id={testId}
      onMouseEnter={() => {
        setIsComponentVisible(true)
      }}
      onMouseLeave={() => {
        setIsComponentVisible(false)
      }}
      tabIndex={0}
      onKeyDown={(e) => {
        if (e.key === 'Tab') {
          setIsComponentVisible(true)
        }
      }}
    >
      {children}
      {isComponentVisible && (
        <ul className="ActionTooltip__actions" style={{ right: rest.right }}>
          {list.map((item: any, i: any) => {
            const canDisplay = typeof item.canDisplay === 'function' ? item.canDisplay(data) : true
            const actionClassnames = cn('wrappper-svg', item.className);
            const controlledClipboardCopyText = data && data[item.copyText] ? data[item.copyText] : data?.uid

            return (
              (canDisplay && item.displayType && item.displayType === 'clipboard') ?

                <li key={i}>
                  {typeof item.label === 'function' ? <RenderLabel component={item.label} data={data} /> : <div className={actionClassnames}>{item.label}</div>}
                </li>
                :
                canDisplay && (
                  item.title !== undefined
                    ? <Tooltip
                      key={i}
                      content={typeof item.title === 'function' ? item.title(data) : item.title}
                      position="top"
                    >
                      <li tabIndex={0}
                        data-test-id="title-action"
                        onBlur={(e) => {
                          if (i === list.length - 1) {
                            setIsComponentVisible(false)
                          }
                        }}
                        onClick={(e) => {
                          setIsComponentVisible(!isComponentVisible)
                          item.action(e, data, rest)
                        }}
                        onKeyDown={(e) => {
                          if (e.key === 'Enter') {
                            setIsComponentVisible(!isComponentVisible)
                            item.action(e, data, rest)
                          }
                        }}
                      >
                        {typeof item.label === 'function' ? <RenderLabel component={item.label} data={data} /> : <div className={actionClassnames} data-test-id="label-tooltip">{item.label}</div>}
                      </li>
                    </Tooltip>
                    : <li tabIndex={0}
                      data-test-id="no-title-action"
                      onBlur={(e) => {
                        if (i === list.length - 1) {
                          setIsComponentVisible(false)
                        }
                      }}
                      onClick={(e) => {
                        setIsComponentVisible(!isComponentVisible)
                        item.action(e, data, rest)
                      }}
                      onKeyDown={(e) => {
                        if (e.key === 'Enter') {
                          setIsComponentVisible(!isComponentVisible)
                          item.action(e, data, rest)
                        }
                      }}
                    >
                      {typeof item.label === 'function' ? <RenderLabel component={item.label} data={data} /> : <div className={actionClassnames} data-test-id="label-tooltip">{item.label}</div>}
                    </li>
                )
            )
          })}
        </ul>
      )}
    </div>
  )
}

ActionTooltip.defaultProps = {
  testId: 'cs-actiontooltip',
  type: 'primary'
} as Partial<IActionTooltip>

export default ActionTooltip