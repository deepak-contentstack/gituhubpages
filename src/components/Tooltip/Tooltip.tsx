import './Tooltip.css'

import React, { createRef } from 'react'

import cn from 'classnames'

import Tippy, { TippyProps } from '@tippyjs/react'

// import Icon from '../Icon2'

enum ArrowDirection {
  "left-start" = "right-start",
  "left-end" = "right-end",
  "left" = "right",

  "right-start" = "left-start",
  "right-end" = "left-end",
  "right" = "left",

  "top-start" = "bottom-start",
  "top-end" = "bottom-end",
  "top" = "bottom",

  "bottom-start" = "top-start",
  "bottom-end" = "top-end",
  "bottom" = "top",
};

enum PaddingClass {
  "left-start" = "pl-10",
  "left-end" = "pl-10",
  "left" = "pl-10",

  "right-start" = "pr-10",
  "right-end" = "pr-10",
  "right" = "pr-10",

  "top-start" = "pt-10",
  "top-end" = "pt-10",
  "top" = "pt-10",

  "bottom-start" = "pb-10",
  "bottom-end" = "pb-10",
  "bottom" = "pb-10",
};
interface ITooltip extends TippyProps {
  content: string | React.ReactNode | { text: string; icon: any }
  position: string
  showArrow?: boolean
  onCreate?: (data: any) => void
  trigger?: string
  onHidden?: () => any
  disabled?: boolean
  variantType?: "dark" | "light" | "menuIcon"
  visible?: boolean,
  showClose?: boolean,
  onClose?: () => void,
  type?: "primary" | "secondary" | "dynamic",
  maxWidth?: any,
  style?: React.CSSProperties,
  [propsKey: string]: any
  testId?: string
  customTooltipCss?: string,
  wrapperElementType?: "div" | "span",
}

const Tooltip: React.FunctionComponent<ITooltip> = (props: any) => {
  const node = createRef<HTMLInputElement>()
  const { content, position, children, isMenuIcon, visible, showClose, testId, style, customTooltipCss, ...rest } = props

  const showArrow = props.showArrow !== undefined ? props.showArrow : false
  const variant = props.variantType || "dark"
  const tooltipType = props.type !== undefined ? props.type : "primary"
  const disabled = props.disabled ? props.disabled : false
  const interactive = typeof content === "string" ? false : true;

  let className =
    variant === "dark"
      ? "tippy-box-dark"
      : tooltipType === "primary"
        ? "tippy-box-light tippy-box-light--primary"
        : `tippy-box-light tippy-box-light--${tooltipType}`
  className = variant === "menuIcon" ? "tippy-box-light tippy-box-light-menu" : className
  className = customTooltipCss ? customTooltipCss : className
  className = cn(className, props.className);
  const wrapperClassNames = cn("tippy-wrapper", { [`${PaddingClass[position]}`]: showArrow });

  const getContent = () => {
    if (variant === "menuIcon") {
      return getKeyIcon(content.text, content.icon);
    }

    if (showArrow) {
      return (
        <div className="flex">
          {content}
          <div className={`tippy-arrow-${variant} tippy-arrow-${variant}-${ArrowDirection[position]}`} />
        </div >
      );
    }

    return content;
  }

  const wrapWithClose = (content) => {
    return (
      <>
        <div className="flex-v-center">
          <div>{content}</div>
          <div className="closeContainer flex-v-center flex-h-center" onClick={props.onClose ? () => props.onClose() : null}>
            {/* <Icon icon='Cancel' /> */}
          </div>
        </div>

      </>
    );
  };

  const WrapperTag = getWrapperType(props.wrapperElementType || "div");
  return (
    <Tippy
      content={showClose ? wrapWithClose(getContent()) : getContent()}
      className={className}
      placement={position}
      arrow={showArrow}
      interactive={interactive}
      disabled={disabled}
      visible={visible}
      {...rest}
    >
      <WrapperTag ref={node} className={wrapperClassNames} data-test-id={testId} style={style}>
        {children}
      </WrapperTag>
    </Tippy>
  )
}

const getKeyIcon = (text: string, icon: any) => {
  return (
    <div className="flex-v-center">
      <div>{text}</div>
      <div className="KeyShortcut">{icon}</div>
    </div>
  )
}
const getWrapperType = (type: string) => {
  if (type === "span") {
    return "span";
  }
  return "div";
}

Tooltip.defaultProps = {
  testId: "cs-tooltip",
  wrapperElementType: "div",
  trigger: "mouseenter focusin"
}

export default Tooltip
