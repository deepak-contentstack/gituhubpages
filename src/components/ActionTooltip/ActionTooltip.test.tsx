import React from 'react'
import renderer from 'react-test-renderer'
import { mount } from 'enzyme'
// import Icon from '../Icon2/Icon'
import { act } from '@testing-library/react'
import ActionTooltip from './ActionTooltip'
// import ClipBoard from '../ClipBoard/ClipBoard'
import Tooltip from '../Tooltip/Tooltip'

describe('ActionTooltip', () => {
  const findByTestAttr = (wrapper, val) => {
    return wrapper.find(`[data-test-id="${val}"]`)
  }

  const toolList = [
    {
      label: <div></div>,
      title: 'Refresh',
      action: () => {
        console.log('Refresh triggered')
      }
    },
    {
      label: <div></div>,
      title: 'Delete',
      action: () => {
        console.log('Delete triggered')
      }
    },
    {
      label: <div></div>,
      title: 'Copy',
      action: () => {
        console.log('Copy triggered')
      }
    }
  ]

  const toolListNoLabelIcon = [
    {
      label: 'Refresh',
      title: 'Refresh',
      action: () => {
        console.log('Refresh triggered')
      }
    }
  ]

  const toolListWithClassname = [
    {
      label: <div></div>,
      title: 'Delete',
      action: () => {
        console.log('Delete triggered')
      },
      className: 'ActionListItem--warning'
    },
    {
      label: <div></div>,
      title: 'Move',
      action: () => {
        console.log('Move triggered')
      },
      className: 'ActionListItem--warning'
    },
    {
      label: <div></div>,
      title: 'Copy UID',
      action: () => {
        console.log('Copy triggered')
      },
      displayType: 'clipboard',
      copyText: 'ct_uid'
    }
  ]

  const noTitle = [
    {
      label: <div></div>,
      // title: 'Refresh',
      action: () => {
        console.log('Refresh triggered')
      }
    }
  ]

  const labelAsFunction = [
    {
      label: () => {
        return <div></div>
      },
      title: 'Refresh',
      action: () => {
        console.log('Refresh triggered')
      }
    }
  ]

  const labelAsFunctionWithoutTitle = [
    {
      label: () => {
        return <div></div>
      },
      action: () => {
        console.log('Refresh triggered')
      }
    }
  ]

  it('render Default ActionTooltip', () => {
    const tree = renderer
      .create(
        <ActionTooltip list={toolList}>
          <div style={{ width: '600px', height: '30px', border: '2px solid grey' }}>Item 1</div>
        </ActionTooltip>
      )
      .toJSON()
    expect(tree).toMatchSnapshot()
  })

  it('Checks if it renders the component correctly', () => {
    const tree = renderer
      .create(
        <ActionTooltip list={toolList}>
          <ActionTooltip className="ActionTooltip-row" list={toolList}>
            <div style={{ width: '600px', height: '30px', border: '2px solid grey' }}>Item 1</div>
          </ActionTooltip>
        </ActionTooltip>
      )
      .toJSON()
    expect(tree).toMatchSnapshot()
  })

  it('Checks if it renders the component correctly with type secondary', () => {
    const tree = renderer
      .create(
        <ActionTooltip list={toolList}>
          <ActionTooltip type="secondary" className="ActionTooltip-row" list={toolList}>
            <div style={{ width: '600px', height: '30px', border: '2px solid grey' }}>Item 1</div>
          </ActionTooltip>
        </ActionTooltip>
      )
      .toJSON()
    expect(tree).toMatchSnapshot()
  })

  it('Checks if it renders the component correctly with custom classnames list', () => {
    const tree = renderer
      .create(
        <ActionTooltip list={toolList}>
          <ActionTooltip type="secondary" className="ActionTooltip-row" list={toolListWithClassname}>
            <div style={{ width: '600px', height: '30px', border: '2px solid grey' }}>Item 1</div>
          </ActionTooltip>
        </ActionTooltip>
      )
      .toJSON()
    expect(tree).toMatchSnapshot()
  })

  it('Checks the tooltip hover', () => {
    const ActionTooltipComponent = mount(
      <ActionTooltip list={toolList} className="ActionTooltip-row" type="secondary">
        Can Augmented And Future Reality Build Empathy?
      </ActionTooltip>
    )
    expect(ActionTooltipComponent).toHaveLength(1)
    const hoverDiv = findByTestAttr(ActionTooltipComponent, 'cs-actiontooltip')
    expect(hoverDiv).toHaveLength(1)
    hoverDiv.simulate('mouseenter')
    expect(ActionTooltipComponent.find(Tooltip)).toHaveLength(3)
    hoverDiv.simulate('mouseleave')
    expect(ActionTooltipComponent.find(Tooltip)).toHaveLength(0)
  })

  it('Checks the clipboard component and tooltip', () => {
    const ActionTooltipComponent = mount(
      <ActionTooltip list={toolListWithClassname} className="ActionTooltip-row" type="secondary">
        Can Augmented And Future Reality Build Empathy?
      </ActionTooltip>
    )
    expect(ActionTooltipComponent).toHaveLength(1)
    const hoverDiv = findByTestAttr(ActionTooltipComponent, 'cs-actiontooltip')
    expect(hoverDiv).toHaveLength(1)
    hoverDiv.simulate('mouseenter')
    //expect(ActionTooltipComponent.find(Tooltip)).toHaveLength(3)
    //expect(ActionTooltipComponent.find(ClipBoard)).toHaveLength(1)
    hoverDiv.simulate('mouseleave')
    expect(ActionTooltipComponent.find(Tooltip)).toHaveLength(0)
    //expect(ActionTooltipComponent.find(ClipBoard)).toHaveLength(0)
  })

  it('Checks the clipboard component and tooltip inner functions', () => {
    const ActionTooltipComponent = mount(
      <ActionTooltip list={toolListWithClassname} className="ActionTooltip-row" type="secondary">
        <div style={{ width: '600px', height: '30px', border: '2px solid grey' }}>Item 1</div>
      </ActionTooltip>
    )
    expect(ActionTooltipComponent).toHaveLength(1)
    const hoverDiv = findByTestAttr(ActionTooltipComponent, 'cs-actiontooltip')
    expect(hoverDiv).toHaveLength(1)
    hoverDiv.simulate('mouseenter')

    //expect(ActionTooltipComponent.find(Tooltip)).toHaveLength(3)
    //expect(ActionTooltipComponent.find(ClipBoard)).toHaveLength(1)
    const toolTip = ActionTooltipComponent.find(Tooltip).first()
    const toolTipRender = findByTestAttr(toolTip, 'title-action')
    expect(toolTipRender).toHaveLength(1)

    toolTipRender.simulate('click')
    expect(ActionTooltipComponent.find(Tooltip)).toHaveLength(0)
    //expect(ActionTooltipComponent.find(ClipBoard)).toHaveLength(0)
  })

  it('Checks the tooltip without title', () => {
    const ActionTooltipComponent = mount(
      // @ts-ignore
      <ActionTooltip list={noTitle} className="ActionTooltip-row" type="secondary">
        <div style={{ width: '600px', height: '30px', border: '2px solid grey' }}>Item 1</div>
      </ActionTooltip>
    )
    expect(ActionTooltipComponent).toHaveLength(1)
    const hoverDiv = findByTestAttr(ActionTooltipComponent, 'cs-actiontooltip')
    expect(hoverDiv).toHaveLength(1)
    hoverDiv.simulate('mouseenter')

    const toolTipRender = findByTestAttr(ActionTooltipComponent, 'no-title-action')
    expect(toolTipRender).toHaveLength(1)

    toolTipRender.simulate('click')
    expect(ActionTooltipComponent.find(Tooltip)).toHaveLength(0)
    //expect(ActionTooltipComponent.find(ClipBoard)).toHaveLength(0)
  })

  it('Checks the event click outside', () => {
    const map = {}
    document.addEventListener = jest.fn((e, cb) => {
      map[e] = cb
    })
    const handleClickOutside = jest.fn()

    const ActionTooltipComponent = mount(
      <ActionTooltip list={toolListWithClassname} className="ActionTooltip-row" type="secondary">
        <div style={{ width: '600px', height: '30px', border: '2px solid grey' }}>Item 1</div>
      </ActionTooltip>
    )

    expect(ActionTooltipComponent).toHaveLength(1)
    const hoverDiv = findByTestAttr(ActionTooltipComponent, 'cs-actiontooltip')
    expect(hoverDiv).toHaveLength(1)
    hoverDiv.simulate('mouseenter')
    //expect(ActionTooltipComponent.find(Tooltip)).toHaveLength(3)
    //expect(ActionTooltipComponent.find(ClipBoard)).toHaveLength(1)

    act(() => {
      //@ts-ignore
      map.click({
        target:
          document.body ||
          document?.activeElement === document?.getElementById('sidebar-extension') ||
          document?.activeElement?.clientWidth == 315
      })
      /* fire events that update state */
    })
  })
  it('triggers the keydown event to close dropdown', () => {
    const map = {}
    document.addEventListener = jest.fn((e, cb) => {
      map[e] = cb
    })
    const handleHideDropdown = jest.fn()

    const ActionTooltipComponent = mount(
      <ActionTooltip list={toolListWithClassname} className="ActionTooltip-row" type="secondary">
        <div style={{ width: '600px', height: '30px', border: '2px solid grey' }}>Item 1</div>
      </ActionTooltip>
    )

    expect(ActionTooltipComponent).toHaveLength(1)
    const hoverDiv = findByTestAttr(ActionTooltipComponent, 'cs-actiontooltip')
    expect(hoverDiv).toHaveLength(1)
    hoverDiv.simulate('mouseenter')
    //expect(ActionTooltipComponent.find(Tooltip)).toHaveLength(3)
    //expect(ActionTooltipComponent.find(ClipBoard)).toHaveLength(1)

    act(() => {
      //@ts-ignore
      map.keydown({ key: 'Escape' })
      /* fire events that update state */
    })
  })

  it('Checks if the Icons are rendering  correctly', () => {
    const ActionTooltipComponent = mount(
      <ActionTooltip list={toolList} className="ActionTooltip-row" type="secondary">
        <div style={{ width: '600px', height: '30px', border: '2px solid grey' }}>Item 1</div>
      </ActionTooltip>
    )

    expect(ActionTooltipComponent).toHaveLength(1)
    const hoverDiv = findByTestAttr(ActionTooltipComponent, 'cs-actiontooltip')
    expect(hoverDiv).toHaveLength(1)
    hoverDiv.simulate('mouseenter')
    //expect(ActionTooltipComponent.find(Icon)).toHaveLength(3)
    expect(ActionTooltipComponent.find(Tooltip)).toHaveLength(3)
  })

  it('Checks if label is rendered if not a component', () => {
    const ActionTooltipComponent = mount(
      <ActionTooltip list={toolListNoLabelIcon} className="ActionTooltip-row" type="secondary">
        <div style={{ width: '600px', height: '30px', border: '2px solid grey' }}>Item 1</div>
      </ActionTooltip>
    )

    expect(ActionTooltipComponent).toHaveLength(1)
    const hoverDiv = findByTestAttr(ActionTooltipComponent, 'cs-actiontooltip')
    expect(hoverDiv).toHaveLength(1)
    hoverDiv.simulate('mouseenter')
    //expect(ActionTooltipComponent.find(Icon)).toHaveLength(0)
    expect(ActionTooltipComponent.find(Tooltip)).toHaveLength(1)
    const toolTip = ActionTooltipComponent.find(Tooltip)
    const labelRender = findByTestAttr(toolTip, 'label-tooltip')
    expect(labelRender).toHaveLength(1)
  })

  it('Checks if label is rendered as a RenderLabel component', () => {
    const ActionTooltipComponent = mount(
      <ActionTooltip list={labelAsFunction} className="ActionTooltip-row" type="secondary">
        <div style={{ width: '600px', height: '30px', border: '2px solid grey' }}>Item 1</div>
      </ActionTooltip>
    )

    expect(ActionTooltipComponent).toHaveLength(1)
    const hoverDiv = findByTestAttr(ActionTooltipComponent, 'cs-actiontooltip')
    expect(hoverDiv).toHaveLength(1)
    hoverDiv.simulate('mouseenter')
    //expect(ActionTooltipComponent.find(Icon)).toHaveLength(1)
    expect(ActionTooltipComponent.find(Tooltip)).toHaveLength(1)
    const toolTip = ActionTooltipComponent.find(Tooltip)
    const checkRenderLabel = toolTip.find('RenderLabel')
    expect(checkRenderLabel).toHaveLength(1)
  })

  it('Checks if label is rendered as a RenderLabel component without title', () => {
    const ActionTooltipComponent = mount(
      //@ts-ignore
      <ActionTooltip list={labelAsFunctionWithoutTitle} className="ActionTooltip-row" type="secondary">
        <div style={{ width: '600px', height: '30px', border: '2px solid grey' }}>Item 1</div>
      </ActionTooltip>
    )

    expect(ActionTooltipComponent).toHaveLength(1)
    const hoverDiv = findByTestAttr(ActionTooltipComponent, 'cs-actiontooltip')
    expect(hoverDiv).toHaveLength(1)
    hoverDiv.simulate('mouseenter')
    const checkRenderLabel = ActionTooltipComponent.find('RenderLabel')
    expect(checkRenderLabel).toHaveLength(1)
  })
})
