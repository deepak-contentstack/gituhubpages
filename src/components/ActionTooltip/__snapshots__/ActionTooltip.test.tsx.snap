// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ActionTooltip Checks if it renders the component correctly 1`] = `
<div
  className="ActionTooltip ActionTooltip--primary"
  data-test-id="cs-actiontooltip"
  onKeyDown={[Function]}
  onMouseEnter={[Function]}
  onMouseLeave={[Function]}
  tabIndex={0}
>
  <div
    className="ActionTooltip ActionTooltip--primary ActionTooltip-row"
    data-test-id="cs-actiontooltip"
    onKeyDown={[Function]}
    onMouseEnter={[Function]}
    onMouseLeave={[Function]}
    tabIndex={0}
  >
    <div
      style={
        Object {
          "border": "2px solid grey",
          "height": "30px",
          "width": "600px",
        }
      }
    >
      Item 1
    </div>
  </div>
</div>
`;

exports[`ActionTooltip Checks if it renders the component correctly with custom classnames list 1`] = `
<div
  className="ActionTooltip ActionTooltip--primary"
  data-test-id="cs-actiontooltip"
  onKeyDown={[Function]}
  onMouseEnter={[Function]}
  onMouseLeave={[Function]}
  tabIndex={0}
>
  <div
    className="ActionTooltip ActionTooltip--secondary ActionTooltip-row"
    data-test-id="cs-actiontooltip"
    onKeyDown={[Function]}
    onMouseEnter={[Function]}
    onMouseLeave={[Function]}
    tabIndex={0}
  >
    <div
      style={
        Object {
          "border": "2px solid grey",
          "height": "30px",
          "width": "600px",
        }
      }
    >
      Item 1
    </div>
  </div>
</div>
`;

exports[`ActionTooltip Checks if it renders the component correctly with type secondary 1`] = `
<div
  className="ActionTooltip ActionTooltip--primary"
  data-test-id="cs-actiontooltip"
  onKeyDown={[Function]}
  onMouseEnter={[Function]}
  onMouseLeave={[Function]}
  tabIndex={0}
>
  <div
    className="ActionTooltip ActionTooltip--secondary ActionTooltip-row"
    data-test-id="cs-actiontooltip"
    onKeyDown={[Function]}
    onMouseEnter={[Function]}
    onMouseLeave={[Function]}
    tabIndex={0}
  >
    <div
      style={
        Object {
          "border": "2px solid grey",
          "height": "30px",
          "width": "600px",
        }
      }
    >
      Item 1
    </div>
  </div>
</div>
`;

exports[`ActionTooltip render Default ActionTooltip 1`] = `
<div
  className="ActionTooltip ActionTooltip--primary"
  data-test-id="cs-actiontooltip"
  onKeyDown={[Function]}
  onMouseEnter={[Function]}
  onMouseLeave={[Function]}
  tabIndex={0}
>
  <div
    style={
      Object {
        "border": "2px solid grey",
        "height": "30px",
        "width": "600px",
      }
    }
  >
    Item 1
  </div>
</div>
`;
